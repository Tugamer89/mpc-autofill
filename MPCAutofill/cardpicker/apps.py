from django.apps import AppConfig


class CardpickerConfig(AppConfig):
    name = "cardpicker"
