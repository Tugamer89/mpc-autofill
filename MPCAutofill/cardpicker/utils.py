import json
import time
from math import floor
from typing import Any

import ratelimit
import requests

TEXT_BOLD = "\033[1m"
TEXT_END = "\033[0m"


def time_to_hours_minutes_seconds(t: float) -> tuple[int, int, int]:
    hours = int(floor(t / 3600))
    mins = int(floor(t / 60) - hours * 60)
    secs = int(t - (mins * 60) - (hours * 3600))
    return hours, mins, secs


def log_hours_minutes_seconds_elapsed(t0: float) -> None:
    hours, mins, secs = time_to_hours_minutes_seconds(time.time() - t0)
    print("Elapsed time: ", end="")
    if hours > 0:
        print(f"{hours} hour{'s' if hours != 1 else ''}, ", end="")
    print(f"{mins} minute{'s' if mins != 1 else ''} and {secs} second{'s' if secs != 1 else ''}.")


@ratelimit.sleep_and_retry  # type: ignore  # `ratelimit` does not implement decorator typing correctly
@ratelimit.limits(calls=1, period=0.1)  # type: ignore  # `ratelimit` does not implement decorator typing correctly
def get_json_endpoint_rate_limited(url: str) -> dict[str, Any]:
    return json.loads(requests.get(url).content)
