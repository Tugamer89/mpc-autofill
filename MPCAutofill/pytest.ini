[pytest]
DJANGO_SETTINGS_MODULE = MPCAutofill.settings
