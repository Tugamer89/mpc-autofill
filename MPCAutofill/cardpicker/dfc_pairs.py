import time

from bulk_sync import bulk_sync

from cardpicker.integrations.integrations import get_configured_game_integration
from cardpicker.models import DFCPair


def sync_dfcs() -> None:
    t0 = time.time()
    game_integration = get_configured_game_integration()
    if game_integration is None:
        print("There is no game configured - no DFC pairs have been imported.")
        return
    print(f"Importing DFC pairs for the game {game_integration.__name__}...")
    dfc_pairs = game_integration.get_dfc_pairs()
    key_fields = ("front",)
    bulk_sync(new_models=dfc_pairs, key_fields=key_fields, filters=None, db_class=DFCPair)
    print(f"Finished importing DFC pairs - this task took {(time.time() - t0):.2f} seconds.")
