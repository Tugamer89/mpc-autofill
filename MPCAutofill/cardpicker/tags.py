"""
Tags for cards.
"""

import re
from typing import Optional

from cardpicker import models
from cardpicker.constants import NSFW


class Tags:
    def __init__(self) -> None:
        self.tags = {
            tag.name.lower(): tag for tag in [models.Tag(name=NSFW, aliases=[], parent=None), *models.Tag.objects.all()]
        }

    def extract_name_and_tags(self, name: Optional[str]) -> tuple[str, set[str]]:
        """
        This function unpacks a folder or image name which contains a name component and some number of tags
        into its constituents.
        Tags are wrapped in either [square brackets] or (parentheses), and any combination of [] and () can be used
        within a single name.
        """

        if not name:
            return "", set()

        tag_set: set[str] = set()
        tag_parts = re.findall(r"\(([^\(\)]+)\)|\[([^\[\]]+)\]", name)  # Get content of () and []
        cleaned_parts = list(map(lambda x: x[0] if len(x[0]) != 0 else x[1], tag_parts))
        name_with_no_tags = name  # tags will be removed from this name below
        for tag_part in cleaned_parts:
            raw_tags = [x.strip() for x in tag_part.split(",")]
            for raw_tag in raw_tags:
                lowercase_tag = raw_tag.lower()

                # identify if this is a valid tag. if it is, add the tag's name to the set
                tag_object: Optional[models.Tag] = None
                if lowercase_tag in self.tags.keys():
                    tag_object = self.tags[lowercase_tag]
                else:
                    for tag in self.tags.values():
                        if lowercase_tag in [alias.lower() for alias in tag.aliases]:
                            tag_object = tag
                            break
                if tag_object is None:
                    continue
                tag_set.add(tag_object.name)

                # `tag_object` also implies all of its parents
                current_tag = tag_object
                while current_tag.parent is not None:
                    tag_set.add(current_tag.parent.name)
                    current_tag = current_tag.parent

                # this is a little ugly. remove all instances of `raw_tag` inside () or [] in the name.
                escaped_raw_tag = re.escape(raw_tag)
                while True:
                    match = re.search(
                        rf"\(.*({escaped_raw_tag},? *).*.*?\)|\[.*({escaped_raw_tag},? *).*.*?\]", name_with_no_tags
                    )
                    if match is None or not any(match.groups()):
                        break
                    for i, group in enumerate(match.groups()):
                        if group is not None:
                            start, end = match.start(i + 1), match.end(i + 1)
                            if start > 0 and end > 0:
                                name_with_no_tags = name_with_no_tags[0:start] + name_with_no_tags[end:]

        artifacts: list[tuple[str, str]] = [  # remove these extra bits from the name
            ("( )", ""),
            ("()", ""),
            ("[ ]", ""),
            ("[]", ""),
            ("[, ", "["),
            (", ]", "]"),
            ("(, ", "("),
            (", )", ")"),
        ]
        for artifact, replacement in artifacts:
            name_with_no_tags = name_with_no_tags.replace(artifact, replacement)
        return name_with_no_tags, tag_set


__all__ = ["Tags"]
