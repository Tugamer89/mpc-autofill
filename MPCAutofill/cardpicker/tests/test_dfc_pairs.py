import pytest

from cardpicker.dfc_pairs import sync_dfcs
from cardpicker.models import DFCPair


class TestSyncDFCs:
    # region fixtures

    @pytest.fixture(autouse=True)
    def autouse_dummy_integration(self, dummy_integration):
        pass

    # endregion

    # region tests

    def test_comprehensive_snapshot(self, snapshot, django_settings):
        sync_dfcs()
        assert list(DFCPair.objects.all().order_by("front")) == snapshot()

    def test_upsert(self, django_settings):
        sync_dfcs()
        pk_to_identifier_1 = {x.pk: (x.front, x.back) for x in DFCPair.objects.all()}
        sync_dfcs()
        pk_to_identifier_2 = {x.pk: (x.front, x.back) for x in DFCPair.objects.all()}
        assert pk_to_identifier_1 == pk_to_identifier_2

    # endregion
