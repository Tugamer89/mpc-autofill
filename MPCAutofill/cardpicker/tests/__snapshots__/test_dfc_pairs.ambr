# name: TestSyncDFCs.test_comprehensive_snapshot
  list([
    <DFCPair: Ratman // Batman>,
    <DFCPair: Ratwoman // Batwoman>,
  ])
# ---
